src/java.base/share/classes/sun/security/util/math/IntegerModuloP.java

```diff
@@ -208,5 +211,134 @@ default ImmutableIntegerModuloP pow(BigInteger b) {
+                        y.setProduct(x);
+                    }
+                    x.setSquare();
+                }
+
+                return y.fixed();
+            }
+        }
+
+        final class Secp256R1 implements MultiplicativeInverser {
+            private static final Secp256R1 instance = new Secp256R1();
+
+            @Override
+            public ImmutableIntegerModuloP inverse(IntegerModuloP imp) {
+                // Invert imp with a modular exponentiation: the modulus is
+                //     p = FFFFFFFF 00000001 00000000 00000000
+                //         00000000 FFFFFFFF FFFFFFFF FFFFFFFF
+                // and the exponent is (p -2).
+                //  p -2 = FFFFFFFF 00000001 00000000 00000000
+                //         00000000 FFFFFFFF FFFFFFFF FFFFFFFD
+                //
+                // There are 4 contiguous 31-bit set, and thus imp^(2^31 - 1)
+                // is pre-computed to speed up the computation.
 
+                // calculate imp ^ (2^31 - 1)
+                MutableIntegerModuloP t = imp.mutable();
+                for (int i = 30; i != 0; i--) {
+                    t.setSquare();
+                    t.setProduct(imp);
+                }
+
+                // calculate (1 / imp)
+                MutableIntegerModuloP d = t.mutable();
+                for (int i = 31; i < 256; i++) {
+                    d.setSquare();
+                    switch (i) {
+                        // For contiguous 31-bit set.
+                        case 190, 221, 252 -> {
+                            d.setProduct(t);
+                        }
+                        // For individual 1-bit set.
+                        case 31, 63, 253, 255 -> {
+                            d.setProduct(imp);
+                        }
+                    }
+                }
+
+                return d.fixed();
+            }
+        }
+
+        final class Secp256R1Field implements MultiplicativeInverser {
+            private static final Secp256R1Field instance = new Secp256R1Field();
+            private static final BigInteger b =
+                    P256OrderField.MODULUS.subtract(BigInteger.TWO);
+            @Override
+            public ImmutableIntegerModuloP inverse(IntegerModuloP imp) {
+                // Invert imp with a modular exponentiation: the modulus is
+                //     n = FFFFFFFF 00000000 FFFFFFFF FFFFFFFF
+                //         BCE6FAAD A7179E84 F3B9CAC2 FC632551
+                // and the exponent is (n -2).
+                //  n - 2 = FFFFFFFF 00000000 FFFFFFFF FFFFFFFF
+                //          BCE6FAAD A7179E84 F3B9CAC2 FC63254F
+                //
+                // There are 3 contiguous 32-bit set, and imp^(2^16 - 1)
+                // is pre-computed to speed up the computation.
+
+                // calculate imp ^ (2^16 - 1)
+                MutableIntegerModuloP t = imp.mutable();
+                for (int i = 15; i != 0; i--) {
```

Comment:
Fun!

You can further reduce the number of multiplications:
t3 = t^2 * t
t15 = t3 ^ 4 * t3
t255 = t15^16 * t15
t65535 = t255^256 * t255
only four multiplications